docs/rfcs/0623-redis-backend.md

```diff
@@ -0,0 +1,237 @@
+
+Files in opendal will be represented as `Hash` in Redis, a kind of semi-structured data type:
+
+```
+                                           ┌───────────────────────────┐
+                                           │  "mode" : "file"          │
+Key: /home/monika/                         ├───────────────────────────┤
+                                           │  "last_modified" :        │
+Key: /home/monika/poem0.txt ─────────────► │  "<rfc3339 timestamp>"    │
+                                           ├───────────────────────────┤
+Key: /home/monika/characters/              │  "content_length" :       │
+                                           │  "224"                    │
+Key: /home/monika/characters/you.chr       ├───────────────────────────┤
+                                           │  "content_md5" :          │
+                                           │  "<md5_string>"           │
+                                           ├───────────────────────────┤
+                                           │  other metadata fields    │
+                                           ├───────────────────────────┤
+                                           │  "content" :              │
+                                           │  "<real content of file>" │
+                                           └───────────────────────────┘
+```
+
+The [redis-rs](https://crates.io/crates/redis)'s high level APIs is preferred.
+
+```rust
+let client = redis::Client::open("redis://localhost:6379")?;
+let con = client.get_async_connection()?;
+```
+
+## Create File
+
+If user is creating a file with root `/home/monika/`, and relative path `poem0.txt`, the `HSET` command of Redis will be called.
+
+```rust
+// mode: ObjectMode
+// path: relative path string
+let path = get_abs_path(path);  // /home/monika/ <> /poem.txt -> /home/monika/poem.txt
+let last_modified = OffsetDatetime::now_utc().to_string();
+
+let kvs = [("content", String::new()), ("last_modified", last_modified), ("content_length", "0".to_string()), ("content_md5", "".to_string())];
+con.hset_multiple(path, &kvs)
+    .await
+    .map_err(|e| other(BackendError::new(..)))?;
+```
+
+```redis
+HSET /home/monika/poem0.txt mode file
+```
+
+This will create a key-value pair, whose key is `/home/monika/poem0.txt`, value is a string to string hashmap containing a tuple `("mode", "file")`.
+
+## Read File
+
+Opendal empowers users to read with the `path` object, `offset` of the cursor and `size` to read. Sadly in redis each time reading the whole object's data will be transmitted, making it opendal's responsibility to move the cursor and to discard the bytes read.
+
+For example, user want to read from poem0.txt. The content of file will be stored in `content` field.
+
+```rust
+// path: "poem0.txt"
+// offset: Option<u64>, the offset of reading
+// size: Option<u64>, the size of reading
+let path = get_abs_path(path);
+let buf: Vec<u8> = con.hget(path, "content")
```

Comment:
So we will read whole content?

Reply:
Sadly, yes...

Reply:
Please take a look at https://github.com/datafuselabs/opendal/pull/623#discussion_r961307554